### DISCUSSION TITLE
Rye's shims break cargo test? (Is this a bug, or have I just broken something ...)

### DISCUSSION THREAD

**Discussion by brettc** [astral-sh/rye#851] [General]
Hi. I'm converting a python/c++ project to a python/rust project, and it has been going great. I'm experienced with python/c++, but a rust noob. I'm using rye / pyo3.rs, working with ndarray / numpy, and it all *just works*. Or did, until today. I'm not even sure what I did to break it. But here is what happens:

* `rye sync` works
* `maturin develop` works fine
* `cargo test` fails with `ld: library 'python3.12' not found`
* same with `cargo bench`
* If I uninstall `rye self uninstall`, then the cargo commands work fine.
* So it seems like it is a problem with the shims being there...?

Note 1: I've uninstalled/reinstalled rust/rye/python and it persists.
Note 2: I have even replicated the problem on another Mac mini m1. So it isn't just my laptop.

Tech deets:
Apple M2 Pro
Sonoma: 14.3.1 (23D60)
rye 0.28.0
Python 3.12.2

Any hints, or revelations of the obvious that I have stupidly overlooked, would be appreciated.

**Comment by bluss** [comment]
This might be issue #646

**Reply by brettc** [reply]
Ah, I see. Further evidence that this might be the problem is that I see:
```
  = note: ld: warning: search path '/install/lib' not found
          ld: library 'python3.12' not found
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
@mitsuhiko mentions something about `/install`.

Is there a easy work-around for this? (I guess I can just have a shell open which does not source the shim code...)

**Reply by bluss** [reply]
Does it work? Put in as `sitecustomize.py` in the relevant venv's site-packages.

```python
def monkey_patch_sysconfig():
    import os
    import sys
    import sysconfig
    sysconfig.get_config_vars()

    patch_dict = sysconfig._CONFIG_VARS
    patch_var = 'LIBDIR'
    old_path = patch_dict[patch_var]
    if old_path.startswith("/install"):
        new_path = os.path.join(sys.base_prefix, patch_dict[patch_var][len("/install/"):])
        print("Patch {}: {} -> {}".format(patch_var, old_path, new_path))
        patch_dict[patch_var] = new_path

monkey_patch_sysconfig()

```

This is not a real solution. Mostly curious.

**Reply by brettc** [reply]
Hi. Thanks for the idea. I tried it, and I can see it is being picked up:
```
~/conrad/wdd2 main*
❯ python -c "import sysconfig; print(sysconfig.get_config_var('LIBDIR'))"

Patch LIBDIR: /install/lib -> /Users/brett/.rye/py/cpython@3.12.2/install/lib
/Users/brett/.rye/py/cpython@3.12.2/install/lib
```

But I get the same error from cargo as before. So I'm not sure how it is searching.
I tried adding the folder above to LD_LIBRARY_PATH and DYLD_FALLBACK_LIBRARY_PATH, but neither worked.

**Reply by bluss** [reply]
It could be that it's looking at the pkgconfig file or so, that also needs to be patched. I think rye can solve this patching eventually, but it's not a super quick fix.

**Reply by gauravkumar37** [reply]
Folks stumbling on this issue and trying out the `monkey_patch_sysconfig` above. It works :rocket: 
However, if you are using it inside cargo, print to stdout is reserved so either comment the print statement or redirect it to stderr
`print(..., file=sys.stderr)`

Or just use https://github.com/bluss/sysconfigpatcher

**Reply by bluss** [reply]
the successor to that patch function is https://github.com/bluss/sysconfigpatcher which is a little bit less hacky (but not much)

**Reply by gauravkumar37** [reply]
Yes, I just edited my comment. The instructions weren't super clear on the repo. Here's what I did:
```sh
$ python3 -c 'import sysconfig; print(sysconfig.get_config_var("LIBDIR"))'
/install/lib

$ uv python find --system
/home/ec2-user/.local/share/uv/python/cpython-3.11.10-linux-x86_64-gnu/bin/python3

$ sysconfigpatcher /home/ec2-user/.local/share/uv/python/cpython-3.11.10-linux-x86_64-gnu

$ python3 -c 'import sysconfig; print(sysconfig.get_config_var("LIBDIR"))'
/home/ec2-user/.local/share/uv/python/cpython-3.11.10-linux-x86_64-gnu/lib
```

**Reply by bluss** [reply]
Great. For convenience there is also `python -m sysconfig | grep LIBDIR`

**Comment by brettc** [comment]
What I cannot figure out is why it worked fine for the last 2 months, and then suddenly broke. I suspect I must have had the "right" python libs available on my path somehow.

**Comment by soddygo** [comment]
"rye self uninstall" ，it useful for me .  
My computer configuration is: MacBook Pro M1 Max. This error has been troubling me for a long time. when i want to use pyo3 .

  my  error info  like  : 
```text
error: linking with `cc` failed: exit status: 1
  |
  = note: env -u IPHONEOS_DEPLOYMENT_TARGET -u TVOS_DEPLOYMENT_TARGET -u XROS_DEPLOYMENT_TARGET LC_ALL="C" PATH="xxxxxxxxx
  = note: ld: warning: search path '/install/lib' not found
          ld: library 'python3.12' not found
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
